network/handlers/drain_test.go

```diff
@@ -55,41 +109,56 @@ func TestDrainMechanics(t *testing.T) {
 		t.Errorf("Probe status = %d, wanted %d", got, want)
 	}
 
-	// Start to drain, and cancel the context when it returns.
+	// Start to drain, and close the channel when it returns.
 	done := make(chan struct{})
 	go func() {
 		defer close(done)
 		drainer.Drain()
 	}()
 
 	select {
-	case <-time.After(40 * time.Millisecond):
-		// Drain is blocking.
 	case <-done:
 		t.Error("Drain terminated prematurely.")
+	case <-init:
+		// OK.
 	}
+	mt.advance(40 * time.Millisecond)
+
 	// Now send a request to reset things.
+	rc := mt.resetCalls
 	drainer.ServeHTTP(w, req)
+	if mt.resetCalls != rc+1 {
+		t.Errorf("ResetCalls = %d, want: %d", mt.resetCalls, rc+1)
+	}
 
 	// Check for 503 as a probe response when shutting down.
 	resp = httptest.NewRecorder()
 	drainer.ServeHTTP(resp, probe)
 	if got, want := resp.Code, http.StatusServiceUnavailable; got != want {
 		t.Errorf("Probe status = %d, wanted %d", got, want)
 	}
+	// Verify no reset was called.
+	if mt.resetCalls != rc+1 {
```

Comment:
I think this would read better if you captured it before the call, even if this is technically correct.

Reply:
but `rc` is captured before the call? Or you want before the initial Drain?